src/preconditions/Cooldown.ts

```diff
@@ -0,0 +1,54 @@
+import { Precondition, PreconditionContext } from '../lib/structures/Precondition';
+import { BucketType } from '../lib/types/Enums';
+import type { Message } from 'discord.js';
+import type { Command } from '../lib/structures/Command';
+import { Bucket } from '@sapphire/ratelimits';
+
+export interface CooldownContext extends PreconditionContext {
+	bucketType?: BucketType;
+	delay?: number;
+	limit?: number;
+}
+
+export class CorePrecondition extends Precondition {
+	public buckets = new WeakMap<Command, Bucket<string>>();
+
+	public run(message: Message, command: Command, context: CooldownContext) {
+		if (!context.delay || context.delay === 0) return this.ok();
+
+		const bucket = this.getBucket(command, context);
+		const remaining = bucket.take(this.getID(message, context));
+
+		return remaining === 0
+			? this.ok()
+			: this.error(
+					this.name,
+					`You have just used this command. Try again in ${Math.ceil(remaining / 1000)} second${remaining > 1000 ? 's' : ''}.`,
+					{ remaining }
+			  );
+	}
+
+	private getID(message: Message, context: CooldownContext) {
+		switch (context.bucketType) {
+			case BucketType.Global:
+				return this.client.user!.id;
```

Comment:
Please prefer `client.id` which is defined by Sapphire to rely less in caching (in case somebody uses a sweeper the wrong way:tm:):
```suggestion
				return this.client.id;
```

Alternatively... why not just
```suggestion
				return 'global';
```

We surely don't need this to be strictly keyed by the bot's ID, unless this is being used for Redis (and sharing it with multiple bots).

Reply:
Resolved in d23b0de